File: src/main/scala/io/github/amerousful/kafka/protocol/KafkaProtocolBuilder.scala
package io.github.amerousful.kafka.protocol

import com.softwaremill.quicklens.ModifyPimp
import io.github.amerousful.kafka.protocol.AutoOffsetReset.AutoOffsetReset
import io.github.amerousful.kafka.protocol.SaslMechanism.SaslMechanism
import org.apache.kafka.clients.consumer.ConsumerConfig._
import org.apache.kafka.clients.producer.ProducerConfig._

import scala.concurrent.duration.FiniteDuration

final case class KafkaProtocolBuilder(kafkaProtocol: KafkaProtocol) {

  def broker(broker: KafkaBroker): KafkaProtocolBuilder = brokers(broker)

  def brokers(brokers: KafkaBroker*): KafkaProtocolBuilder = {
    val bootstrapConfig: String = brokers.map(b => b.host + ":" + b.port).mkString(",")

    addProducerProperty(org.apache.kafka.clients.producer.ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapConfig)
      .addConsumerProperty(org.apache.kafka.clients.consumer.ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapConfig)
  }

  def acks(acks: String): KafkaProtocolBuilder = addProducerProperty(ACKS_CONFIG, acks)

  def producerKeySerializer(serializer: String): KafkaProtocolBuilder =
    addProducerProperty(KEY_SERIALIZER_CLASS_CONFIG, serializer)

  def producerValueSerializer(serializer: String): KafkaProtocolBuilder =
    addProducerProperty(VALUE_SERIALIZER_CLASS_CONFIG, serializer)

  def consumerKeyDeserializer(deserializer: String): KafkaProtocolBuilder =
    addConsumerProperty(KEY_DESERIALIZER_CLASS_CONFIG, deserializer)

  def consumerValueDeserializer(deserializer: String): KafkaProtocolBuilder =
    addConsumerProperty(VALUE_DESERIALIZER_CLASS_CONFIG, deserializer)

  def producerIdenticalSerializer(serializer: String): KafkaProtocolBuilder =
    producerKeySerializer(serializer)
      .producerValueSerializer(serializer)

  def consumerIdenticalDeserializer(deserializer: String): KafkaProtocolBuilder =
    consumerKeyDeserializer(deserializer)
      .consumerValueDeserializer(deserializer)

  def credentials(username: String, password: String, sslEnabled: Boolean, saslMechanism: SaslMechanism): KafkaProtocolBuilder = {
    val protocol = sslEnabled match {
      case true => "SASL_SSL"
      case false => "SASL_PLAINTEXT"
    }
    val jaasConfigModule = saslMechanism match {
      case SaslMechanism.plain => "plain.PlainLoginModule"
      case SaslMechanism.scram_sha_256 => "scram.ScramLoginModule"
      case SaslMechanism.scram_sha_512 => "scram.ScramLoginModule"
    }
    addConsumerAndProducerProperty("security.protocol", protocol)
      .addConsumerAndProducerProperty("sasl.mechanism", saslMechanism.toString)
      .addConsumerAndProducerProperty("sasl.jaas.config",
        s"""org.apache.kafka.common.security.$jaasConfigModule required username="$username" password="$password";""".stripMargin)
  }

  def ssl(protocol: String, keystoreLocation: String, keystorePassword: String, keyPassword: String): KafkaProtocolBuilder = {
    addConsumerAndProducerProperty("security.protocol", protocol)
      .addConsumerAndProducerProperty("ssl.keystore.location", keystoreLocation)
      .addConsumerAndProducerProperty("ssl.keystore.password", keystorePassword)
      .addConsumerAndProducerProperty("ssl.key.password", keyPassword)
  }

  def ssl(
           protocol: String,
           keystoreLocation: String,
           keystorePassword: String,
           keyPassword: String,
           trustStoreLocation: String,
           trustStorePassword: String
         ): KafkaProtocolBuilder = {
    ssl(protocol, keystoreLocation, keystorePassword, keyPassword)
      .addConsumerAndProducerProperty("ssl.truststore.location", trustStoreLocation)
      .addConsumerAndProducerProperty("ssl.truststore.password", trustStorePassword)
  }

  def addProducerProperty(key: String, value: String): KafkaProtocolBuilder = {
    this.modify(_.kafkaProtocol.producerProperties)(_ + (key -> value))
  }

  def addConsumerProperty(key: String, value: AnyRef): KafkaProtocolBuilder = {
    this.modify(_.kafkaProtocol.consumerProperties)(_ + (key -> value))
  }

  private def addConsumerAndProducerProperty(key: String, value: String): KafkaProtocolBuilder =
    addProducerProperty(key, value)
      .addConsumerProperty(key, value)

  def messageMatcher(matcher: KafkaMatcher): KafkaProtocolBuilder =
    this.modify(_.kafkaProtocol.messageMatcher).setTo(matcher)

  def matchByKey(): KafkaProtocolBuilder = messageMatcher(KafkaKeyMatcher)

  def matchByValue(): KafkaProtocolBuilder = messageMatcher(KafkaValueMatcher)

  def replyTimeout(timeout: FiniteDuration): KafkaProtocolBuilder = this.modify(_.kafkaProtocol.replyTimeout).setTo(Some(timeout))

  def replyConsumerName(name: String): KafkaProtocolBuilder = addConsumerProperty(GROUP_ID_CONFIG, name)

  def schemaUrl(url: String): KafkaProtocolBuilder = addConsumerAndProducerProperty("schema.registry.url", url)

  def disableAutoCommit() = addConsumerProperty(ENABLE_AUTO_COMMIT_CONFIG, "false")

  def autoOffsetResetPolicy(value: AutoOffsetReset) = addConsumerProperty(AUTO_OFFSET_RESET_CONFIG, value)

  def build: KafkaProtocol = kafkaProtocol
}

object KafkaProtocolBuilder {
  implicit def toKafkaProtocol(builder: KafkaProtocolBuilder): KafkaProtocol = builder.build

  val Default: KafkaProtocolBuilder = KafkaProtocolBuilder(KafkaProtocol.apply())
}

File: src/main/java/io/github/amerousful/kafka/javaapi/KafkaProtocolBuilder.java
package io.github.amerousful.kafka.javaapi;

import edu.umd.cs.findbugs.annotations.NonNull;
import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;
import io.github.amerousful.kafka.protocol.javaapi.KafkaMessageMatchers;
import scala.Enumeration;

import java.time.Duration;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static io.gatling.javaapi.core.internal.Converters.toScalaDuration;
import static io.gatling.javaapi.core.internal.Converters.toScalaSeq;

public final class KafkaProtocolBuilder implements ProtocolBuilder {

    private final io.github.amerousful.kafka.protocol.KafkaProtocolBuilder wrapped;

    KafkaProtocolBuilder(io.github.amerousful.kafka.protocol.KafkaProtocolBuilder wrapped) {
        this.wrapped = wrapped;
    }

    @NonNull
    public KafkaProtocolBuilder broker(KafkaBroker broker) {
        return new KafkaProtocolBuilder(wrapped.broker(broker.asScala()));
    }

    @NonNull
    public KafkaProtocolBuilder brokers(KafkaBroker... brokers) {
        List<io.github.amerousful.kafka.protocol.KafkaBroker> listOfBrokersAsScala =
                Arrays.stream(brokers).map(KafkaBroker::asScala).collect(Collectors.toList());
        return new KafkaProtocolBuilder(wrapped.brokers(toScalaSeq(listOfBrokersAsScala)));
    }

    @NonNull
    public KafkaProtocolBuilder acks(@NonNull String acks) {
        return new KafkaProtocolBuilder(wrapped.acks(acks));
    }

    @NonNull
    public KafkaProtocolBuilder producerKeySerializer(@NonNull String serializer) {
        return new KafkaProtocolBuilder(wrapped.producerKeySerializer(serializer));
    }

    @NonNull
    public KafkaProtocolBuilder producerValueSerializer(@NonNull String serializer) {
        return new KafkaProtocolBuilder(wrapped.producerValueSerializer(serializer));
    }

    @NonNull
    public KafkaProtocolBuilder consumerKeyDeserializer(@NonNull String deserializer) {
        return new KafkaProtocolBuilder(wrapped.consumerKeyDeserializer(deserializer));
    }

    @NonNull
    public KafkaProtocolBuilder consumerValueDeserializer(@NonNull String deserializer) {
        return new KafkaProtocolBuilder(wrapped.consumerValueDeserializer(deserializer));
    }

    @NonNull
    public KafkaProtocolBuilder consumerIdenticalDeserializer(@NonNull String deserializer) {
        return consumerKeyDeserializer(deserializer)
                .consumerValueDeserializer(deserializer);
    }

    @NonNull
    public KafkaProtocolBuilder producerIdenticalSerializer(@NonNull String serializer) {
        return new KafkaProtocolBuilder(wrapped.producerIdenticalSerializer(serializer));
    }

    @NonNull
    public KafkaProtocolBuilder addProducerProperty(@NonNull String key, @NonNull String value) {
        return new KafkaProtocolBuilder(wrapped.addProducerProperty(key, value));
    }

    @NonNull
    public KafkaProtocolBuilder addConsumerProperty(@NonNull String key, @NonNull String value) {
        return new KafkaProtocolBuilder(wrapped.addConsumerProperty(key, value));
    }

    @NonNull
    public KafkaProtocolBuilder credentials(@NonNull String username, @NonNull String password, boolean sslEnabled, @NonNull Enumeration.Value saslMechanism) {
        return new KafkaProtocolBuilder(wrapped.credentials(username, password, sslEnabled, saslMechanism));
    }

    @NonNull
    public KafkaProtocolBuilder replyTimeout(long timeout) {
        return replyTimeout(Duration.ofSeconds(timeout));
    }

    @NonNull
    public KafkaProtocolBuilder replyTimeout(@NonNull Duration timeout) {
        return new KafkaProtocolBuilder(wrapped.replyTimeout(toScalaDuration(timeout)));
    }

    @NonNull
    public KafkaProtocolBuilder messageMatcher(@NonNull KafkaMessageMatcher matcher) {
        return new KafkaProtocolBuilder(wrapped.messageMatcher(KafkaMessageMatchers.toScala(matcher)));
    }

    @NonNull
    public KafkaProtocolBuilder matchByKey() {
        return new KafkaProtocolBuilder(wrapped.matchByKey());
    }

    @NonNull
    public KafkaProtocolBuilder matchByValue() {
        return new KafkaProtocolBuilder(wrapped.matchByValue());
    }

    @NonNull
    public KafkaProtocolBuilder replyConsumerName(@NonNull String name) {
        return new KafkaProtocolBuilder(wrapped.replyConsumerName(name));
    }

    @Override
    public Protocol protocol() {
        return wrapped.build();
    }
}
